spec/stream_queue_spec.cr

```diff
@@ -178,4 +210,179 @@ describe LavinMQ::StreamQueue do
+      queue_name = Random::Secure.hex
+      consumer_tag = Random::Secure.hex
+      offset = 3
+
+      StreamQueueSpecHelpers.publish(queue_name, offset + 1)
+
+      offset.times { StreamQueueSpecHelpers.consume_one(queue_name, consumer_tag) }
+      sleep 0.1
+
+      # consume again, should start from last offset automatically
+      msg = StreamQueueSpecHelpers.consume_one(queue_name, consumer_tag)
+      StreamQueueSpecHelpers.offset_from_headers(msg.properties.headers).should eq offset + 1
+    end
+
+    it "reads offsets from file on init" do
+      queue_name = Random::Secure.hex
+      vhost = Server.vhosts["/"]
+      offsets = [84_i64, 24_i64, 1_i64, 100_i64, 42_i64]
+      tag_prefix = "ctag-"
+      StreamQueueSpecHelpers.publish(queue_name, 1)
+
+      data_dir = File.join(vhost.data_dir, Digest::SHA1.hexdigest queue_name)
+      msg_store = LavinMQ::StreamQueue::StreamQueueMessageStore.new(data_dir, nil)
+      offsets.each_with_index do |offset, i|
+        msg_store.update_consumer_offset(tag_prefix + i.to_s, offset)
+      end
+      msg_store.close
+      sleep 0.1
+
+      msg_store = LavinMQ::StreamQueue::StreamQueueMessageStore.new(data_dir, nil)
+      offsets.each_with_index do |offset, i|
+        msg_store.last_offset_by_consumer_tag(tag_prefix + i.to_s).should eq offset
+      end
+      msg_store.close
+    end
+
+    it "only saves one entry per consumer tag" do
+      queue_name = Random::Secure.hex
+      vhost = Server.vhosts["/"]
+      offsets = [84_i64, 24_i64, 1_i64, 100_i64, 42_i64]
+      consumer_tag = "ctag-1"
+      StreamQueueSpecHelpers.publish(queue_name, 1)
+
+      data_dir = File.join(vhost.data_dir, Digest::SHA1.hexdigest queue_name)
+      msg_store = LavinMQ::StreamQueue::StreamQueueMessageStore.new(data_dir, nil)
+      offsets.each do |offset|
+        msg_store.update_consumer_offset(consumer_tag, offset)
+      end
+      msg_store.close
+      sleep 0.1
+
+      msg_store = LavinMQ::StreamQueue::StreamQueueMessageStore.new(data_dir, nil)
+      msg_store.last_offset_by_consumer_tag(consumer_tag).should eq offsets.last
+      msg_store.@consumer_offsets.size.should eq 15
+
+      msg_store.close
+    end
+
+    it "does not track offset if x-stream-offset is set" do
+      queue_name = Random::Secure.hex
+      consumer_tag = Random::Secure.hex
+      c_args = AMQP::Client::Arguments.new({"x-stream-offset": 0})
+
+      StreamQueueSpecHelpers.publish(queue_name, 2)
+      msg = StreamQueueSpecHelpers.consume_one(queue_name, consumer_tag, c_args)
+      StreamQueueSpecHelpers.offset_from_headers(msg.properties.headers).should eq 1
+      sleep 0.1
+
+      # should consume the same message again since tracking was not saved from last consume
+      msg_2 = StreamQueueSpecHelpers.consume_one(queue_name, consumer_tag)
+      StreamQueueSpecHelpers.offset_from_headers(msg_2.properties.headers).should eq 1
+    end
+
+    it "should not use saved offset if x-stream-offset is set" do
```

Comment:
Do we want this behavior?  I'm thinking that if you setup a consumer that says `x-stream-offset=10` and it start consumes to `offset=20` if I restart that consumer then I would expect it to continue from 21, not 10, even though the `x-stream-offset` says 10. 
Otherwise I have to specify `x-stream-offset` first time and then somehow remove the argument if the consumer is restarted and should continue. WDYT?

Reply:
It's a bit tricky IMO. But in general I think that if you specify x-stream-offset when consuming, that should take precedent over other things. 

Let's say you get messages 10-20, and then decide you want to get them again for whatever reason (or jump to another part of the stream), there would be no way for you to do that if we use our tracking. 

An alternative could be to use an extra argument (x-stream-auto-offset or something), where we think of x-stream-offset as a starting point, and then use our internal tracking for any subsequent requests.

Reply:
How this works in Kafka is that offsets are stored based on consumer group name, so if you want to start over using a new offset you need to rename the group. Changing offset when using a group that has already a stored offset on the broker has no effect.  They do have an API call into kafka to reset a consumer group offset so you can tell a running consumer to start reading from somewhere else, but that is just a side note. 

For rabbitmq Streams it looks like the offset it stored based on consumer name (same feature where you can group  consumers on) so if order to "reset" the offset and use another value you need to rename the consumer.  Do we have a name of something for a consumer that we can use in this case?

Reply:
> For rabbitmq Streams it looks like the offset it stored based on consumer name (same feature where you can group consumers on) so if order to "reset" the offset and use another value you need to rename the consumer. Do we have a name of something for a consumer that we can use in this case?

We use the consumer tag, so you can always "start over" by just setting a new tag.

Reply:
Okay, then if we can reset by changing consumer tag I think the broker should use stored offsets if there are any and if not use the one given by the consumer using the argument `x-stream-offset` 

So use stored offset if exists for consumer tags, otherwise use the offset given by consumer

Reply:
Hmm, I'm not sure I agree - I think that if you provide `x-stream-offset` you explicitly say what you want. That would also  change existing behavior where you always get what you request when providing an `x-stream-offset`. But let's discuss more during tomorrows meeting!

Reply:
@snichme added `x-stream-use-automatic-offset` so we can have it both ways :)

Reply:
yay, nice!